core/ast_node.js

```diff
@@ -0,0 +1,657 @@
+ * @private
+ */
+Blockly.ASTNode.prototype.findNextEditableField_ = function(location,
+    parentInput, opt_first) {
+  var fieldRow = parentInput.fieldRow;
+  var fieldIdx = fieldRow.indexOf(location);
+  var startIdx = opt_first ? 0 : fieldIdx + 1;
+  var astNode;
+  for (var i = startIdx; i < fieldRow.length; i++) {
+    var field = fieldRow[i];
+    if (field.isCurrentlyEditable()) {
+      astNode = Blockly.ASTNode.createFieldNode(field);
+      break;
+    }
+  }
+  return astNode;
+};
+
+/**
+ * Get either the previous editable field, or get the first editable field for
+ * the given input.
+ * @param {!Blockly.Field} location The current location of the cursor.
+ * @param {!Blockly.Input} parentInput The parentInput of the field.
+ * @param {?Boolean} opt_last If true find the last editable field otherwise get
+ *     the previous field.
+ * @return {?Blockly.ASTNode} The ast node holding the previous or last field or
+ *     null if no next field exists.
+ * @private
+ */
+Blockly.ASTNode.prototype.findPreviousEditableField_ = function(location,
+    parentInput, opt_last) {
+  var fieldRow = parentInput.fieldRow;
+  var fieldIdx = fieldRow.indexOf(location);
+  var previousField = null;
+  var startIdx = opt_last ? fieldRow.length - 1 : fieldIdx - 1;
+  for (var i = startIdx; i >= 0; i--) {
+    var field = fieldRow[i];
+    if (field.isCurrentlyEditable()) {
+      previousField = field;
+      break;
+    }
+  }
+  return Blockly.ASTNode.createFieldNode(previousField);
+};
+
+/**
+ * Get the first field or connection that is either editable or has connection
+ * value of not null.
+ * @param {!Blockly.Connection} location Current location in the ast.
+ * @param {!Blockly.Input} parentInput The parent input of the field or
+ *     connection.
+ * @return {?Blockly.ASTNode} The ast node holding the next field or connection.
+ * @private
+ */
+Blockly.ASTNode.prototype.findNextForInput_ = function(location, parentInput){
+  var inputs = location.sourceBlock_.inputList;
+  var curIdx = inputs.indexOf(parentInput);
+  if (curIdx <= -1) {return;}
+  var nxtIdx = curIdx + 1;
+  var newAstNode = null;
+
+  for (var i = nxtIdx; i < inputs.length; i++) {
+    var newInput = inputs[i];
+    var field = this.findNextEditableField_(location, newInput, true);
+    if (field) {
+      newAstNode = Blockly.ASTNode.createFieldNode(field);
+      break;
+    } else if (newInput.connection) {
+      var connection = newInput.connection;
+      newAstNode = Blockly.ASTNode.createConnectionNode(connection);
+      break;
+    }
+  }
+  return newAstNode;
+};
+
+/**
+ * Find the next input or field given a field location.
+ * @param {!Blockly.Field} location Current location in the ast.
+ * @param {!Blockly.Input} parentInput The parent input of the field.
+ * @return {?Blockly.ASTNode} The ast node holding the next field or connection.
+ * @private
+ */
+Blockly.ASTNode.prototype.findNextForField_ = function(location, parentInput) {
```

Comment:
This should be combined with findNextForInput since in both cases you need to iterate through fields and inputs until you find a valid location or hit the end of the block. Right now your code will go from an input to a field, then get stuck on the last field in that input.

You should probably add a test that iterates across a multi-input multi-field block using next and previous to make sure all editable elements are reachable.